key_index.lua

```diff
@@ -0,0 +1,97 @@
+-- Storage to keep track of used keys. Allows to atomically create, delete
+-- and list keys. The keys are synchronized between nginx workers
+-- using ngx.shared.dict. The whole purpose of this module is to avoid
+-- using ngx.shared.dict:get_keys (see https://github.com/openresty/lua-nginx-module#ngxshareddictget_keys),
+-- which blocks all workers and therefore it shouldn't be used with large
+-- amounts of keys.
+
+local KeyIndex = {}
+KeyIndex.__index = KeyIndex
+
+local lock_lib = require("lock")
+
+function KeyIndex.new(shared_dict)
+  local self = setmetatable({}, KeyIndex)
+  self.dict = shared_dict
+  self.keys = {}
+  self.index = {}
+  self.lock = lock_lib.new("__lock_keys", self.dict)
+  return self
+end
+
+-- Loads new keys that might have been added by other workers since last sync.
+function KeyIndex:sync()
+  local N = self.dict:get("__key_count") or 0
+  -- Only sync if there are some new keys.
+  if N ~= #self.keys then
```

Comment:
Actually, `#{1,2,3,nil,4}` is 5 ;-)

But you are right, it will have to be stored separately, because if we delete the last item, then the `#` operator would return lower value.

Reply:
it will be 5 for puc-lua but 3 for luajit, but anyway i guess we are on the same page : )

Reply:
Oh, I forgot this might be different in different implementations :blush: As I said, I'm still a newbie.

Should be fixed in b1d51cd907c0aea0a752ee6ff6481f9ef3405054